pages/DSNP/Announcements.md

```diff
@@ -40,6 +40,16 @@ Additional duplicate Announcements MUST be rejected or ignored.
 Announcements may not be deleted, but some may be marked as invalid by using a [Tombstone Announcement](Types/Tombstone.md), or updated by using an [Update Announcement](Types/Update.md).
 For example, if a user creates a Reaction Announcement, they may remove that reaction by creating a Tombstone Announcement.
 
+## Operations
+
+Each of the these MUST have verifiable approval of the acting
+principal via signature proof or other system.
+
+| Operation | Optional? | Principal | Inputs | Outputs/State Changes |
+|---------- |---------- |---------- |------- |-------------- |
+| Publish Announcement | no | User OR Delegate | User's DSNP Identifier, Announcement Type, Announcement Data | Announcement published to the network |
+| Publish Batch | no | User OR Delegate | User or Delegate's DSNP Identifier, Announcement Type, [Batch Publication File](BatchPublications.md) | Batch announcement published to network |
```

Comment:
My thoughts here...

> Should publish batch be a thing at all?

I'd like to consider this as two questions:
a) Should batching be part of the DSNP spec (as opposed to an implementation-specific spec for Frequency)?
b) If batching _is_ part of the DSNP spec, should it be an optional operation?

If batching is not in scope for DSNP, then we should move all the details of Parquet file structure and so on into the Frequency spec instead. I was proceeding from a basis of trying to capture what is already there in the DSNP spec docs. But I agree it's worth asking the question.

It is true that any batch announcement can be disaggregated into a series of individual announcements. An operational distinction is that all announcements in a batch must be from DSNP users who share a delegation relationship with the principal invoking the batch publish operation. However, once this relationship has been verified for each announcement, the fact that the delegate was the principal does not need to be retained as part of the state, and the state changes are equivalent.

On the other hand, I think we need to provide some clarity for _read_ operations on DSNP. If we define an observer role as one that sees all state change activity, then we can define the state change event stream in an abstract form. If we say batches are an implementation detail, then an observer (i.e. a reader who is outside the boundaries of the implementation and relying on the spec operations it provides) can be defined to see individual announcement events only. If, however, we want to put the responsibility of reading batch files onto the observer, then the batch announcement becomes its own event on the event stream, and the format of the batch file needs to be specified on the DSNP side.

Working it through this way, I would like to propose the following:
1. We document a formal definition of a DSNP implementation as a distributed ledger, with an observable stream of records. Note that we don't need to specify the binary representation of these records (i.e. what it looks like on a blockchain) as long as we can specify the types of records and the data elements that must be observable. In other words, it must be possible to define an output translation function/layer that is able to produce a stream of normalized DSNP records for outside parties to interact with.
2. We specify individual announcements as part of the observable stream.
3. We DO NOT specify batches as part of the observable stream.
4. We move the definitions of batch file structure and handling out of the DSNP spec and make these the domain of the Frequency implementation team.

Reply:
Hmm... It feels odd for batching to be separate from DSNP as it is a core piece to making it work on any current scaled implementation. Or perhaps a better way to put it is that it doesn't feel right that it is part of the Frequency on DSNP spec. I suppose it could be a "separate" spec (like Activity Content).

The responsibility of reading batch files is on the observer. We can try to say that it is merely part of the implementation, but that is more arbitrary than most of the lines. It also "removes" the value of using the bloom filters to skip processing batch files that are not important to your application.

The event stream in theory is nothing but Announcements and it is designed to be consumed in that way, but it does reduce the power of the Batches.

Reply:
Since Frequency does not actually batch certain types of messages, in a way it is optional in that the Frequency implementation uses schema which explicitly say what messages are batched and what aren't.

Maybe that part should be part of the spec, what _may_ be batched and what _must not_ be batched.  Does it make sense to say specifically, e.g., you can't take identifier operations and graph messages out of the main/shared DSNP message stream, because that, it seems to me, undermines important goals of the protocol, like discoverability and verifiable delegation**?

I don't see a batch file author DSNP identifier in the [batch publication spec](https://spec.dsnp.org/DSNP/BatchPublications.html) by the way, there is just a reference that it should be able to be verified.

Reply:
@shannonwells I would turn this on its head somewhat and say that *because* Frequency has made some design decisions for efficiency, it (the implementation) needed to make certain choices about which announcement types are batched or not.

I feel that as long as the announcement operations end up emitting records (events) that reference the announcements or batches and give users a way to retrieve them, discoverability should be OK.

Perhaps what we need to do is specify that certain announcement types must have their delegation verified by the implementation vs. by the reader.

Reply:
I'm going to move that last sentence to a discussion topic and resolve this thread.